solr/core/src/java/org/apache/solr/storage/DirectBufferPool.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.storage;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+public class DirectBufferPool {
+  private final ConcurrentLinkedQueue<ByteBuffer> pool = new ConcurrentLinkedQueue<>();
+  private final int size;
+  private final int blockSize;
+  private final int allocateChunkSize;
+
+  public DirectBufferPool(int size, int blockSize, int allocatChunkSize) {
+    this.size = size;
+    this.blockSize = blockSize;
+    this.allocateChunkSize = allocatChunkSize;
+    if (size % blockSize != 0) {
+      throw new IllegalArgumentException("block size incompatible with size");
+    }
+  }
+
+  public ByteBuffer get() {
```

Comment:
Any concern with performance issues here as the pool gets populated initially given the synchronized block or is the expectation that'll be efficient enough to not cause a problem?

Reply:
Good question; the thought here is that the pool should not need to be terribly large, so any potential bottleneck during ramp-up should be transient and manageable. The point of synchronizing at the top level here is to prevent lower-level attempts (by glibc) to reduce memory contention via MALLOC_ARENA, which could result in memory fragmentation, etc.